torch/quantization/ns/graph_matcher.py

```diff
@@ -224,73 +296,99 @@ def get_matching_node_pairs(
+          (conv_0, conv_0),  # (start_node_a, end_node_a)
+          (qconv_0, qconv_0),  # (start_node_b, end_node_b)
         ),
     }
 
+    Or, if we have a fusion pattern,
+
+    graph_a: x0 -> linear_0 -> relu_0 -> obs_0 -> x1
+             w  -/
+             b  -/
+
+    graph_b: x0 -> quant_0 -> linear_relu_0 -> dequant_0 -> x1
+           packed_params_0 -/
+
+    This function will return the following result:
+    {
+        'linear_relu_0': (  # the name of the node in graph_b
+          (linear_0, relu_0),  # (start_node_a, end_node_a)
+          (linear_relu_0, linear_relu_0),  # (start_node_b, end_node_b)
+        ),
+    }
     """
     non_matchable_functions = get_non_matchable_functions()
     non_matchable_modules = get_non_matchable_modules()
-    graph_a_iterator = _NSGraphMatchableNodesIterator(
+    graph_a_iterator = _NSGraphMatchableSubgraphsIterator(
         gm_a, non_matchable_functions, non_matchable_modules)
-    graph_b_iterator = _NSGraphMatchableNodesIterator(
+    graph_b_iterator = _NSGraphMatchableSubgraphsIterator(
         gm_b, non_matchable_functions, non_matchable_modules)
     results = {}
     type_a_related_to_b = get_type_a_related_to_b()
 
     while True:
         # fetch the next nodes from a and b
-        cur_node_a, cur_node_b = None, None
+        cur_start_node_a, cur_start_node_b = None, None
+        cur_end_node_a, cur_end_node_b = None, None
         try:
-            cur_node_a = next(graph_a_iterator)
+            cur_start_node_a, cur_end_node_a = next(graph_a_iterator)
         except StopIteration:
             pass
         try:
-            cur_node_b = next(graph_b_iterator)
+            cur_start_node_b, cur_end_node_b = next(graph_b_iterator)
         except StopIteration:
             pass
 
         # look up types of a and b for useful error messages
         type_a, type_b = None, None
-        if cur_node_a is not None:
-            type_a = _get_node_target_type(cur_node_a, gm_a)
-        if cur_node_b is not None:
-            type_b = _get_node_target_type(cur_node_b, gm_b)
+        if cur_end_node_a is not None:
+            type_a = _get_node_target_type(cur_end_node_a, gm_a)
+        if cur_end_node_b is not None:
+            type_b = _get_node_target_type(cur_end_node_b, gm_b)
 
         # check for results and determine what to do next
-        if cur_node_a is not None and cur_node_b is not None:
+        if cur_end_node_a is not None and cur_end_node_b is not None:
+            assert isinstance(cur_start_node_a, Node)
+            assert isinstance(cur_start_node_b, Node)
             # both nodes were fetched, check for relatedness
-            if not _node_a_related_to_b(cur_node_a, cur_node_b,
+            # note: relatedness is checked on the start node, i.e.
+            # if a linear-relu pattern is checked, we would check for relatedness
+            # of the linear
+            if not _node_a_related_to_b(cur_start_node_a, cur_start_node_b,
```

Comment:
Since subgraph matching is introduced to graph matcher, would it be natural to extend the relatedness checking to support subgraph also, which might be more general?

Reply:
Sure, that could make sense.  Is there a use case in mind?

Since this is a private API, the eng cost of changing this in the future would be low.

Reply:
Yeah we can think about it in later PR. I don't have specific use case in mind, just thinking if we can relate for example the subgraph (linear, relu) to linear_relu directly, then it would be self-explanatory and we may not need to note "relatedness is checked on the start node".